improvement(container-runtime): Simplify dirty state management in ContainerRuntime

markfields opened the pull request:
## Description

Fixes [AB#34161](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/34161)

Simplifies dirty state tracking by computing it consistently every time its value may need to be updated.

This fixes a bug in Staging Mode where if upon discarding staged changes the only remaining pending op were "non-dirtyable", we would still call the container dirty.  See tests in documentDirty.spec.ts

markfields committed:
Prep

- Introduce currentDirtyState()
- Some no-op changes to simplify the code
- Some other small changes for this PR that won't affect subsequent

markfields committed:
Update each updateDocumentDirtyState callsite to use currentDirtyState()

And add temp notes explaining why the old hardcoded values match the computed value

markfields committed:
Remove updateDocumentDirtyState param

markfields committed:
Remove the asserts from updateDocumentDirtyState

They're provably true now that we're using currentDirtyState rather than taking an arg

These asserts give even more confidence to this change

markfields committed:
Rename only: dirtyContainer to lastEmittedDirty

markfields committed:
Update isDirty to compute it, not use last emitted.

This will only change the result of public isDirty
for those moments between when a change to attach or pending state happens
and when we call updateDocumentDirtyState.

I checked all these codepaths, and they are all synchronous,
so this change will be a no-op for callers of this public API

markfields committed:
Remove redundant if checks before updateDocumentDirtyState

In each case:
- The if check is equivalent to "no longer dirty" (meaning, we would skip if dirty)
- So the question is*: Could we have just switched from saved to dirty?  No.
  - Processing an op can't make us dirty from saved
  - Attaching can't make us dirty from saved

*This is the case where this commit's change would result in an extra dirty event where there isn't one today.

markfields committed:
Consolidate two updateDocumentDirtyState calls

markfields committed:
Final clean up

markfields committed:
revert .only

anthony-murphy commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2528,17 +2528,13 @@ export class ContainerRuntime
-		const oldState = this.dirtyContainer;
-		this.dirtyContainer = false;
-
+		// Replaying is an internal operation and we don't want to generate noise while doing it.
```
do we really need this? in the updateDocumentDirtyState we only emit if the state changes, won't the state just go from dirty to dirty, so nothing will be emitted?

anthony-murphy replied:
ideally we could get rid of emitDirtyDocumentEvent too

markfields replied:
Hmmm I would like to get rid of this too, and you make a good point.

If it's not dirty now, this will be a big no-op.  And the calls to `updateDocumentDirtyState` which can transition to "saved" can't be reached while replaying pending ops.

So, yeah look safe.

markfields replied:
I think I am going to do that in a follow-up PR, it feels more bold of a "refactor" than I'm going for in this one.

anthony-murphy replied:
seems fine, but current change does a bunch of refactoring already.

markfields committed:
Fix unit tests

markfields commented:
Oops - I missed the point of isContainerMessageDirtyable, which is that some messages will sit in PSM but not make the container dirty.  Thankfully a bunch of tests are failing.

Couple different ideas for how to fix it:
- Add an override flag that indicates that the only pending messages are not "dirtyable"
- Flip `isDirty` back to returning the class field (rename `lastEmittedDirty` back to something else).  Leave the rest of the change where `update` fn doesn't take a param, etc.

markfields committed:
Hack in previous behavior around non-dirtyable messages.

The more correct fix is to compute dirty state based on isDirtyable logic every time

markfields committed:
Revert "Hack in previous behavior around non-dirtyable messages."

This reverts commit 44ea39fa492539df4d569191f5d6b64e0c731851.

markfields committed:
Initial try at checking for dirtyable messages every time

anthony-murphy commented:
> Oops - I missed the point of isContainerMessageDirtyable, which is that some messages will sit in PSM but not make the container dirty. Thankfully a bunch of tests are failing.
> 
> Couple different ideas for how to fix it:
> 
> * Add an override flag that indicates that the only pending messages are not "dirtyable"
> * Flip `isDirty` back to returning the class field (rename `lastEmittedDirty` back to something else).  Leave the rest of the change where `update` fn doesn't take a param, etc.
> * Actually inspect the messages in Outbox/PSM inside isDirty call. This is the most correct, and I think even fixes bugs in the current code (where other calls to update don't consider non-dirtyable messages). Shouldn't be costly perf-wise since in most cases the first message we check is dirtyable

why do we not dirty the container on some ops? Does this prevent switching from read to write connection mode or something?

markfields committed:
Try accounting for non-dirtyable messages in isDirty calculation

It seems tests still fail

markfields committed:
Reset state to 49ff2fe82cbf736b8452dbdc67903e581dae319a

Right before updating isDirty to calculate

markfields commented:
Back to Draft state. Might actually abandon, trying 1 or two more things to keep the current behavior but make it slightly more clear/consistent

markfields committed:
Revert change to throw when entering SM if detached

markfields committed:
Clean up - new less-changy change

markfields committed:
Remove .only

markfields committed:
Remove redundant if checks

markfields committed:
Consolidate two updateDocumentDirtyState calls

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2550,14 +2545,13 @@ export class ContainerRuntime
-		// Officially transition from the old state to the new state.
-		this.updateDocumentDirtyState(newState);
+		// This will emit an event if the state changed relative to before replay
+		this.updateDocumentDirtyState();
```
Potential behavior change here (in corner cases): This will recompute dirty state rather than using the "last emitted" state from during the replay.

So this will be another case where if the only pending message is non-dirtyable, the state will say "dirty" when it could/should say "saved".  Thinking through non-dirtyable ops, I don't think this is an issue. Possibly with Agent Scheduler, I'm not familiar with that as much.

But worst case would be a spurious "unsaved changes" dialog after successful reconnect, until the op roundtrips.  Summarizer also checks isDirty state, but Summarizer closes on reconnect so this path isn't a concern.

markfields replied:
Fixed - Updated the "dirty" computation to account for these all the time.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -3261,10 +3254,7 @@ export class ContainerRuntime
-						if (this.dirtyContainer !== checkpointDirtyState) {
-							this.updateDocumentDirtyState(checkpointDirtyState);
-						}
+						this.updateDocumentDirtyState();
```
This is another place where we recompute rather than restoring the previous last-emitted.  Won't affect summarizer (we don't use Order Sequentially there), so should be ok.

markfields replied:
Fixed - Updated the "dirty" computation to account for these all the time.

markfields committed:
Revert one "if" removal, and add comments

markfields committed:
Some ID Allocation op fixes

markfields commented:
Test failures are due to a bug with submitting ID Allocation op before replaying pending states _but we don't schedule a flush_.  So this PR makes the container dirty if that ID Allocation op is the only op, but then it can get stuck that way.

This prompted me to revive #24545 which delays the submission of that op until we're submitting other ops, which solves this issue.  Blocking this PR until that one goes in.

markfields committed:
Check for dirtyable messages every time

markfields committed:
Fix missing check for Immediate mode

markfields committed:
Fix UTs

markfields committed:
Revert ID Allocation "fix" in replay path

It shouldn't be necessary anymore since the dirty state will be correct
since we're computing it robustly.

markfields committed:
Some unit tests

markfields committed:
Cleanup

markfields committed:
More cleanup

markfields committed:
Revert unrelated change

markfields committed:
PSM unit tests

markfields committed:
Add some more tests

markfields committed:
Merge branch 'main' into cr/dirty-saved
